corehq/apps/data_interfaces/models.py

```diff
@@ -1125,8 +1157,81 @@ def properties_fit_definition(self, case_properties):
+        if not case_matching_rule_exists_in_es(case, rule):
+            # If the case isn't found, any duplicate information is unreliable
+            # Just queue this specific record up manually via resave_case
+            resave_case(rule.domain, case, send_post_save_signal=False)
+            return CaseRuleActionResult(num_updates=0)
+
+        try:
+            existing_duplicate = CaseDuplicateNew.objects.get(case_id=case.case_id, action=self)
+        except CaseDuplicateNew.DoesNotExist:
+            existing_duplicate = None
+
+        current_hash = CaseDuplicateNew.case_and_action_to_hash(case, self)
+
+        # Check if the new parameters have changed.
+        if existing_duplicate and existing_duplicate.match_values == current_hash:
+            # Nothing has changed. We can stop processing here
+            return CaseRuleActionResult(num_updates=0)
+
+        with transaction.atomic():
+            if existing_duplicate:
+                existing_duplicate.delete()
+            duplicate_ids = self._create_duplicates(case, rule, current_hash)
+
+        if process_updates:
+            num_updates = self._update_duplicates(duplicate_ids, case, rule)
+            return CaseRuleActionResult(num_updates=num_updates)
+        else:
+            return CaseRuleActionResult(num_updates=0)
+
+    def _create_duplicates(self, case, rule, current_hash):
+        # Handle whether or not this gets inserted as a new duplicate
+        matching_ids = find_matching_case_ids_in_es(case, rule, limit=3)
+
+        other_duplicate_ids = set([case_id for case_id in matching_ids if case_id != case.case_id])
+        if len(other_duplicate_ids) == 0:
+            # This isn't a duplicate, just return
+            return []
+
+        duplicates = [CaseDuplicateNew(case_id=case.case_id, action=self, match_values=current_hash)]
+        missing_ids = self._get_case_ids_not_recorded_as_duplicates(other_duplicate_ids)
+        if missing_ids:
+            # create a new duplicate for anything that currently isn't registered as one
+            new_duplicates = [
+                CaseDuplicateNew(case_id=missing_id, action=self, match_values=current_hash)
+                for missing_id in missing_ids
+            ]
+            duplicates.extend(new_duplicates)
+
+        CaseDuplicateNew.objects.bulk_create(duplicates)
+        return [duplicate.case_id for duplicate in duplicates]
+
+    def _get_case_ids_not_recorded_as_duplicates(self, all_ids):
+        existing_ids = set(CaseDuplicateNew.objects.filter(
+            action=self, case_id__in=all_ids
+        ).values_list('case_id', flat=True))
+        return all_ids - existing_ids
+
+    def _update_duplicates(self, duplicate_ids, case, rule):
+        num_updates = 0
+        if duplicate_ids and self.properties_to_update:
+            num_updates = len(duplicate_ids)
+            num_updates = self._update_cases(case.domain, rule, duplicate_ids)
```

Comment:
Is this expected to update all cases that are duplicates of the current case being processed? Am I understanding correctly that `duplicate_ids` will contain a maximum of 3 cases[^1], an may not be an exhaustive list of duplicates. Does it matter if a subset of duplicates are processed?

[^1]: see the [first line](https://github.com/dimagi/commcare-hq/pull/33859/files#diff-f76d77f2d2a18ebdcf7519f3a15e2156dafaf61d5201f5e506faf40344b00aecR1197) of `_create_duplicates` above